### ISSUE TITLE
FRI rounds vs. committed codewords count

### ISSUE THREAD

**Issue by raugfer** [aszepieniec/stark-anatomy#32]
Hey, I am studying your FRI implementation.

It produces `r` codewords for  `r` rounds with an implied minimum of `1` round. 
https://github.com/aszepieniec/stark-anatomy/blob/fce084f66df77b5ce726cafadb1d82144ebf4790/code/fri.py#L20

But I believe it should produce `r + 1` codewords for `r` rounds.

This is particularly noticeable when looking at the `sample_indices()` call during the `proof()` method, where `codewords[1]`  will trigger an IndexError when `num_rounds() == 1`, given that in this scenario `commit()` produces only `1` entry:
https://github.com/aszepieniec/stark-anatomy/blob/fce084f66df77b5ce726cafadb1d82144ebf4790/code/fri.py#L122

I think the simple fix is just modifying the associated commit/verify loops to iterate one extra step.

However it has the obvious effects of an extra iteration. Therefore I wonder if one would also need to adjust 'num_rounds()' to account for the change (possibly returning one less than originally) and properly accommodate the provided FRI parameters.

Would you be able to clarify that? Thanks for making this tutorial available!

**Comment by aszepieniec** [comment]
Hi @raugfer, thanks for the remark. I'm not sure I agree though -- or maybe I misunderstand entirely, in which case please elaborate.

From reading `Fri.prove` again, I observe that `r+1` codewords *are* being collected into the variable `codewords`. Specifically, the lines

```python
            # collect codeword
            codewords += [codeword]
```

is executed `r` times and below the loop there is an additional increment:


```python
        # collect last codeword too
        codewords = codewords + [codeword]
```

**Comment by aszepieniec** [comment]
You're absolutely right. And I'm not sure whether fixing this discrepancy in one place will cause errors in another. How about you make a PR fixing the issue in the way you think is best, and we address the issues that arise from there (if any)?

**Comment by raugfer** [comment]
I have created the PR #33. Explanation is in the PR text. I think the proper fix will not make changes to `num_rounds()`, but that is exactly what I am not sure an prompted me to open this issue.

**Comment by aszepieniec** [comment]
Thanks for the PR. It's giving me a lot to think about. I'm still trying to wrap my head around where things might break down.

In the mean time, could it be that the right fix is something stupidly simple like setting the index to 0 instead of 1?
```
        top_level_indices = self.sample_indices(proof_stream.prover_fiat_shamir(), len(codewords[0]), len(codewords[-1]), self.num_colinearity_tests)
```

So `len(codewords[0])` is the length of the very first, unfolded, codeword. We want to be selecting pairs of indices that point in there, spaced apart by exactly half the codeword length. And since we are computing the counterpart without modular reduction, we want to sample the index from `range(0,len(codewords[0])/2)`.

So maybe including the `/2` is even better:
```
        top_level_indices = self.sample_indices(proof_stream.prover_fiat_shamir(), len(codewords[0])/2, len(codewords[-1]), self.num_colinearity_tests)
```

Furthermore, the parameter `reduced_size` which takes the value `len(codewords[-1])` is used only to guarantee that after folding `num_rounds` many times, the indices are all still unique. In fact, this filtering turns out to be unnecessary as the security analysis already accounts for that. (But addressing this requires are rather more substantial refactor of the code.)

**Comment by aszepieniec** [comment]
Fixed in 6c5e197.
